src/include/miopen/env.hpp

```diff
@@ -31,127 +31,181 @@
-    };
+template <typename T>
+struct ParseEnvVal
+{
+};
 
-/*
- * Returns false if a feature-controlling environment variable is defined
- * and set to something which disables a feature.
- */
-inline bool IsEnvvarValueDisabled(const char* name)
+template <>
+struct ParseEnvVal<bool>
 {
-    // NOLINTNEXTLINE (concurrency-mt-unsafe)
-    const auto value_env_p = std::getenv(name);
-    if(value_env_p == nullptr)
-        return false;
-    else
+    static bool go(const char* vp)
     {
-        std::string value_env_str = value_env_p;
+        std::string value_env_str{vp};
+
         for(auto& c : value_env_str)
         {
             if(std::isalpha(c) != 0)
             {
                 c = std::tolower(static_cast<unsigned char>(c));
             }
         }
-        return (std::strcmp(value_env_str.c_str(), "disable") == 0 ||
-                std::strcmp(value_env_str.c_str(), "disabled") == 0 ||
-                std::strcmp(value_env_str.c_str(), "0") == 0 ||
-                std::strcmp(value_env_str.c_str(), "no") == 0 ||
-                std::strcmp(value_env_str.c_str(), "off") == 0 ||
-                std::strcmp(value_env_str.c_str(), "false") == 0);
-    }
-}
 
-inline bool IsEnvvarValueEnabled(const char* name)
-{
-    // NOLINTNEXTLINE (concurrency-mt-unsafe)
-    const auto value_env_p = std::getenv(name);
-    if(value_env_p == nullptr)
-        return false;
-    else
-    {
-        std::string value_env_str = value_env_p;
-        for(auto& c : value_env_str)
+        if(std::strcmp(value_env_str.c_str(), "disable") == 0 ||
+           std::strcmp(value_env_str.c_str(), "disabled") == 0 ||
+           std::strcmp(value_env_str.c_str(), "0") == 0 ||
+           std::strcmp(value_env_str.c_str(), "no") == 0 ||
+           std::strcmp(value_env_str.c_str(), "off") == 0 ||
+           std::strcmp(value_env_str.c_str(), "false") == 0)
         {
-            if(std::isalpha(c) != 0)
-            {
-                c = std::tolower(static_cast<unsigned char>(c));
-            }
+            return false;
         }
-        return (std::strcmp(value_env_str.c_str(), "enable") == 0 ||
+        else if(std::strcmp(value_env_str.c_str(), "enable") == 0 ||
                 std::strcmp(value_env_str.c_str(), "enabled") == 0 ||
                 std::strcmp(value_env_str.c_str(), "1") == 0 ||
                 std::strcmp(value_env_str.c_str(), "yes") == 0 ||
                 std::strcmp(value_env_str.c_str(), "on") == 0 ||
-                std::strcmp(value_env_str.c_str(), "true") == 0);
+                std::strcmp(value_env_str.c_str(), "true") == 0)
+        {
+            return true;
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusInvalidValue, "Invalid value for env variable");
+        }
+
+        return false; // shouldn't reach here
     }
-}
+};
+
+template <>
+struct ParseEnvVal<uint64_t>
+{
+    static uint64_t go(const char* vp) { return std::strtoull(vp, nullptr, 0); }
```

Comment:
I suggest we check the result to make sure the parsing was successful to avoid any silent issues where the user thinks they have set the correct value for the env var however, it fails to parse.

Reply:
strtoull is going to return 0 if it fails to parse a number
I'm not sure it's a huge issue. When the int result is 0, we could strcmp to "0" and throw an error if they aren't equal.

Reply:
@cderb we should strcmp to `"0x<zero-seq>"`, `"<zero-seq>"`, where `<zero-seq>` is an arbitrary number of `'0'` characters, because strings like "0000" and "0x0000000000" are valid zeroes.

Reply:
@cderb @JehandadKhan This is example of detection of stroull failure:
```C++
char* endp = nullptr;
const auto value = strtoull(vp, &endp, 0);
if (endp == vp        // No digits found
    || errno != 0     // Underflow/overflow or some other error
    || *endp != '\0') // Some characters remain (can't be parsed)
{
    if (errno !=0) { errno = 0; } // clean up error state
    // Handle error (e.g. return error code to the caller)
}
```